Source/WebCore/platform/graphics/FontMetrics.h

```diff
@@ -21,159 +21,165 @@
+    std::optional<float> lineSpacing() const { return m_lineSpacing; }
+    int intLineSpacing() const { return m_intLineSpacing; }
     void setLineSpacing(float lineSpacing)
     {
-        m_floatLineSpacing = lineSpacing;
+        m_lineSpacing = lineSpacing;
         m_intLineSpacing = lroundf(lineSpacing);
     }
 
-    float xHeight() const { return m_xHeight; }
+    std::optional<float> xHeight() const { return m_xHeight; }
     void setXHeight(float xHeight) { m_xHeight = xHeight; }
-    bool hasXHeight() const { return m_xHeight > 0; }
-    
-    bool hasCapHeight() const { return m_floatCapHeight > 0; }
-    float floatCapHeight() const { return m_floatCapHeight; }
+
+    std::optional<float> capHeight() const { return m_capHeight; }
+    int intCapHeight() const { return m_intCapHeight; }
     void setCapHeight(float capHeight)
     {
-        m_floatCapHeight = capHeight;
+        m_capHeight = capHeight;
         m_intCapHeight = lroundf(capHeight);
     }
-    
-    // Integer variants of certain metrics, used for HTML rendering.
-    int ascent(FontBaseline baselineType = AlphabeticBaseline) const
-    {
-        if (baselineType == AlphabeticBaseline)
-            return m_intAscent;
-        return height() - height() / 2;
-    }
-    
-    int descent(FontBaseline baselineType = AlphabeticBaseline) const
-    {
-        if (baselineType == AlphabeticBaseline)
-            return m_intDescent;
-        return height() / 2;
-    }
-
-    int height(FontBaseline baselineType = AlphabeticBaseline) const
-    {
-        return ascent(baselineType) + descent(baselineType);
-    }
-
-    int lineGap() const { return m_intLineGap; }
-    int lineSpacing() const { return m_intLineSpacing; }
-    
-    int capHeight() const { return m_intCapHeight; }
-    
-    bool hasIdenticalAscentDescentAndLineGap(const FontMetrics& other) const
-    {
-        return ascent() == other.ascent() && descent() == other.descent() && lineGap() == other.lineGap();
-    }
 
     std::optional<float> zeroWidth() const { return m_zeroWidth; }
     void setZeroWidth(float zeroWidth) { m_zeroWidth = zeroWidth; }
 
-    float ideogramWidth() const { return m_ideogramWidth; }
+    std::optional<float> ideogramWidth() const { return m_ideogramWidth; }
     void setIdeogramWidth(float ideogramWidth) { m_ideogramWidth = ideogramWidth; }
 
-    float underlinePosition() const { return m_underlinePosition; }
+    std::optional<float> underlinePosition() const { return m_underlinePosition; }
     void setUnderlinePosition(float underlinePosition) { m_underlinePosition = underlinePosition; }
 
-    float underlineThickness() const { return m_underlineThickness; }
+    std::optional<float> underlineThickness() const { return m_underlineThickness; }
     void setUnderlineThickness(float underlineThickness) { m_underlineThickness = underlineThickness; }
 
+    bool hasIdenticalAscentDescentAndLineGap(const FontMetrics& other) const
```

Comment:
> This function hides the fact that “identical” actually means “same after rounding to int”. So it might also need a name change. Also unclear why it’s a member function.

`hasSameAscentDescentAndLineGapAfterRounding()` or `hasSameIntAscentDescentAndLineGap()`? If we have no better name, I would like to deal with it later in a separate bug.